src/python/pants/engine/internals/uuid.py

```diff
@@ -1,28 +1,49 @@
 # Copyright 2020 Pants project contributors (see CONTRIBUTORS.md).
 # Licensed under the Apache License, Version 2.0 (see LICENSE).
 
-import random
 import uuid
-from dataclasses import dataclass, field
+from dataclasses import dataclass
+from enum import Enum
+from typing import Optional, cast
 
-from pants.engine.rules import _uncacheable_rule, collect_rules
+from pants.engine.rules import collect_rules, rule
+from pants.util.meta import frozen_after_init
 
 
-@dataclass(frozen=True)
+class UUIDScope(Enum):
+    PER_CALL = "call"
+    PER_SESSION = "session"
+
+
+@frozen_after_init
+@dataclass(unsafe_hash=True)
 class UUIDRequest:
-    randomizer: float = field(default_factory=random.random)
+    @staticmethod
+    def _to_scope_name(scope: UUIDScope) -> str:
+        if scope == UUIDScope.PER_CALL:
+            return uuid.uuid4().hex
+        return cast(str, scope.value)
+
+    @classmethod
+    def scoped(cls, scope: UUIDScope) -> "UUIDRequest":
+        return cls(cls._to_scope_name(scope))
```

Comment:
Nit: I think with dataclasses, the convention is to put the fields like `scope: str` at the very top.

Reply:
It definitely is and has been grating. I can stick to convention though.

Reply:
Fixed.